frame/executive/src/lib.rs

```diff
@@ -1073,4 +1073,42 @@ mod tests {
 			assert_eq!(sp_io::storage::get(CUSTOM_ON_RUNTIME_KEY).unwrap(), true.encode());
 		});
 	}
+
+	#[test]
+	fn all_weights_are_recorded_correctly() {
+		new_test_ext(1).execute_with(|| {
+			// Make sure `on_runtime_upgrade` is called for maximum complexity
+			RUNTIME_VERSION.with(|v| *v.borrow_mut() = sp_version::RuntimeVersion {
+				spec_version: 1,
+				..Default::default()
+			});
+
+			let block_number = 1;
+
+			Executive::initialize_block(&Header::new(
+				block_number,
+				H256::default(),
+				H256::default(),
+				[69u8; 32].into(),
+				Digest::default(),
+			));
+
+			// All weights that show up in the `initialize_block_impl`
+			let custom_runtime_upgrade_weight = CustomOnRuntimeUpgrade::on_runtime_upgrade();
+			let runtime_upgrade_weight = <AllModules as OnRuntimeUpgrade>::on_runtime_upgrade();
+			let on_initialize = <AllModules as OnInitialize<u64>>::on_initialize(block_number);
+			let base_block_weight = <Runtime as frame_system::Trait>::BlockExecutionWeight::get();
+			let frame_system_initialize_weight = <frame_system::Module::<Runtime> as OnRuntimeUpgrade>::on_runtime_upgrade();
+
+			// Weights are recorded correctly
+			assert_eq!(
+				frame_system::Module::<Runtime>::block_weight().total(),
+				custom_runtime_upgrade_weight +
+				runtime_upgrade_weight +
+				on_initialize +
+				base_block_weight +
+				frame_system_initialize_weight,
+			);
```

Comment:
```suggestion
			let frame_system_upgrade_weight = <frame_system::Module::<Runtime> as OnRuntimeUpgrade>::on_runtime_upgrade();

			// Weights are recorded correctly
			assert_eq!(
				frame_system::Module::<Runtime>::block_weight().total(),
				custom_runtime_upgrade_weight +
				runtime_upgrade_weight +
				on_initialize +
				base_block_weight +
				frame_system_upgrade_weight,
			);
```

Reply:
Also I think `frame_system::on_initialize` weight is missing

Reply:
There is no frame_system on_initialize, but indeed i added the logic for good measure :)